panic on signal during startup

Still investigating if the kibana dashboard setup is related or just a convenient way to reproduce this.  Reliably getting a panic on ctrl-c immediately after start up using:

```
$ ./apm-server -e -E setup.kibana.host=localhost:5601 -E setup.dashboards.enabled=true
...
2018-01-17T16:51:38.291-0500	INFO	[monitoring]	log/log.go:95	Starting metrics logging every 30s
2018-01-17T16:51:38.292-0500	INFO	elasticsearch/client.go:145	Elasticsearch url: http://localhost:9200
2018-01-17T16:51:38.298-0500	INFO	elasticsearch/client.go:689	Connected to Elasticsearch version 6.2.0
2018-01-17T16:51:38.298-0500	INFO	kibana/client.go:69	Kibana url: http://localhost:5601
^C
2018-01-17T16:51:38.736-0500	INFO	beater/beater.go:72	stopping apm-server...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x6c pc=0x42aad44]
goroutine 25 [running]:
net/http.(*Server).Shutdown(0x0, 0x52a4280, 0xc4202ba180, 0x0, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/net/http/server.go:2488 +0x54
github.com/elastic/apm-server/beater.stop(0x0, 0x12a05f200)
	/go/src/github.com/elastic/apm-server/beater/server.go:51 +0x9b
github.com/elastic/apm-server/beater.(*beater).Stop(0xc42014e500)
	/go/src/github.com/elastic/apm-server/beater/beater.go:73 +0x6f
github.com/elastic/apm-server/vendor/github.com/elastic/beats/libbeat/beat.(Beater).Stop-fm()
	/go/src/github.com/elastic/apm-server/vendor/github.com/elastic/beats/libbeat/cmd/instance/beat.go:288 +0x2f
sync.(*Once).Do(0xc420158250, 0xc4201d2140)
	/usr/local/Cellar/go/1.9.2/libexec/src/sync/once.go:44 +0xbe
github.com/elastic/apm-server/vendor/github.com/elastic/beats/libbeat/service.HandleSignals.func1(0xc420394300, 0xc420158250, 0xc4201d2140)
	/go/src/github.com/elastic/apm-server/vendor/github.com/elastic/beats/libbeat/service/service.go:34 +0x99
created by github.com/elastic/apm-server/vendor/github.com/elastic/beats/libbeat/service.HandleSignals
	/go/src/github.com/elastic/apm-server/vendor/github.com/elastic/beats/libbeat/service/service.go:31 +0x106
```

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -20,7 +20,8 @@ https://github.com/elastic/apm-server/compare/71df0d96445df35afe27f38bcf734a0828
 ==== Bug fixes
 - Updated systemd doc url {pull}354[354]
 - Updated readme doc urls {pull}356[356]
-- Use updated stack trace frame values for calculating error `grouping_keys` {pull}485[485] 
+- Use updated stack trace frame values for calculating error `grouping_keys` {pull}485[485]
+- Fix panic when a SIGINT is delivered before the server is instantiated {pull}580[580]
 
 ==== Added
 - Include build time and revision in version information {pull}396[396]

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -21,7 +21,7 @@ https://github.com/elastic/apm-server/compare/71df0d96445df35afe27f38bcf734a0828
 - Updated systemd doc url {pull}354[354]
 - Updated readme doc urls {pull}356[356]
 - Use updated stack trace frame values for calculating error `grouping_keys` {pull}485[485]
-- Fix panic when a SIGINT is delivered before the server is instantiated {pull}580[580]
+- Fix panic when a signal is delivered before the server is instantiated {pull}580[580]
 
 ==== Added
 - Include build time and revision in version information {pull}396[396]

--- a/beater/beater.go
+++ b/beater/beater.go
@@ -7,14 +7,18 @@ import (
 	"net/http"
 	"regexp"
 
+	"sync"
+
 	"github.com/elastic/beats/libbeat/beat"
 	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/logp"
 )
 
 type beater struct {
-	config *Config
-	server *http.Server
+	config  *Config
+	server  *http.Server
+	mutex   sync.RWMutex
+	stopped bool
 }
 
 // Creates beater
@@ -36,7 +40,8 @@ func New(b *beat.Beat, ucfg *common.Config) (beat.Beater, error) {
 	}
 
 	bt := &beater{
-		config: beaterConfig,
+		config:  beaterConfig,
+		stopped: false,
 	}
 	return bt, nil
 }
@@ -57,7 +62,14 @@ func (bt *beater) Run(b *beat.Beat) error {
 	}
 	go notifyListening(bt.config, pub.Send)
 
+	bt.mutex.Lock()
+	if bt.stopped {
+		defer bt.mutex.Unlock()
+		return nil
+	}
+
 	bt.server = newServer(bt.config, pub.Send)
+	bt.mutex.Unlock()
 
 	err = run(bt.server, lis, bt.config)
 	if err == http.ErrServerClosed {
@@ -70,5 +82,10 @@ func (bt *beater) Run(b *beat.Beat) error {
 // Graceful shutdown
 func (bt *beater) Stop() {
 	logp.Info("stopping apm-server...")
-	stop(bt.server, bt.config.ShutdownTimeout)
+	bt.mutex.Lock()
+	if bt.server != nil {
+		stop(bt.server, bt.config.ShutdownTimeout)
+	}
+	bt.stopped = true
+	bt.mutex.Unlock()
 }

--- a/beater/beater.go
+++ b/beater/beater.go
@@ -6,7 +6,6 @@ import (
 	"net"
 	"net/http"
 	"regexp"
-
 	"sync"
 
 	"github.com/elastic/beats/libbeat/beat"
@@ -16,8 +15,8 @@ import (
 
 type beater struct {
 	config  *Config
+	mutex   sync.RWMutex // guards server and stopped
 	server  *http.Server
-	mutex   sync.RWMutex
 	stopped bool
 }
 

--- a/beater/beater.go
+++ b/beater/beater.go
@@ -15,7 +15,7 @@ import (
 
 type beater struct {
 	config  *Config
-	mutex   sync.RWMutex // guards server and stopped
+	mutex   sync.Mutex // guards server and stopped
 	server  *http.Server
 	stopped bool
 }

